Welcome home refresh (#27)
datastax/astra-streaming-docs: modules/developing/pages/astream-functions.adoc

Functions are implemented using Apache Pulsar(R) functions. See https://pulsar.apache.org/docs/en/functions-overview/[Pulsar Functions overview] for more information about Pulsar functions.

== Add functions in dashboard

Add functions in the Functions tab of the Astra Streaming dashboard.

. Select *Create Function* to get started.
. Choose your function's name and namespace.
+
image::astream-name-function.png[Function and Namespace]

> . Choose the file you want to pull the function from and which function you want to use within that file.

Reviewer: ```suggestion
. Select the file you want to pull the function from and which function you want to use within that file.
```

Astra generates a list of acceptable classes. Python and Java functions are added a little differently from each other.

Python functions are added by loading a Python file (.py) or a zipped Python file (.zip). When adding Python files, the Class Name is specified as the name of the Python file without the extension plus the class you want to execute.

> For example, if the Python file is called `testfunction.py` and the class is `ExclamationFunction`, then the class name would be `testfunction.ExclamationFunction`. The file can contain multiple classes, but only one will be used. If there is no class in the Python file (when using a basic function, for example) just specify the filename without the extension (ex. `function`).

Reviewer: ```suggestion
For example, if the Python file is called `testfunction.py` and the class is `ExclamationFunction`, then the class name is `testfunction.ExclamationFunction`. The file can contain multiple classes, but only one is used. If there is no class in the Python file (when using a basic function, for example), specify the filename without the extension (ex. `function`).
```

+
If the function is set up and ready to accept messages, you should see "Created Successfully!"

. Use `./pulsar-admin functions list --tenant <tenant-name>` to list the functions in your tenant and confirm your new function was created.

== Testing Your Function

> Triggering a function is a convenient way to test that the function is working. When you trigger a function, you are publishing a message on the function’s input topic, which triggers the function to run. If the function has an output topic and the function returns data to the output topic, it will be displayed.

Reviewer: ```suggestion
Triggering a function is a convenient way to test that the function is working. When you trigger a function, you are publishing a message on the function’s input topic, which triggers the function to run. If the function has an output topic and the function returns data to the output topic, that data is displayed.
```

If you need to update any other setting of the function, delete and then re-add the function.

To update your function, select your function in the *Functions* dashboard.

image::astream-function-update.png[Update Function]

. Select *Change File* to find your function locally and click *Open*.

. Update your function's *Instances* and *Timeout*. When you're done, click *Update*.

> . An *Updates Submitted Successfully* flag will appear to let you know your function has been updated.

Reviewer: Is this a flag or an alert? I've heard the rectangle, green notation that appears at the top of a user's screen only referred to as an alert.

== Deleting Your Function

To delete a function, select the function to be deleted in the *Functions* dashboard.

image::astream-delete-function.png[Delete Function]

. Click *Delete*.

> . A popup will ask you to confirm deletion by entering the function's name and clicking *Delete*.

Reviewer: ```suggestion
. A popup asks you to confirm deletion by entering the function's name and clicking *Delete*.
```
